### ISSUE TITLE
2019/09

### ISSUE THREAD

**Issue by XiangyunHuang** [XiangyunHuang/r-bloggers#23]
- [ ] 模型选择的一些基本思想和方法 https://cosx.org/2015/08/some-basic-ideas-and-methods-of-model-selection
- [ ] Bandit 算法与推荐系统 https://cosx.org/2017/05/bandit-and-recommender-systems/
- [ ] 机器学习与数据挖掘 https://cs.nju.edu.cn/zhouzh/zhouzh.files/publication/cccf07.pdf
- [ ] 互联网征信中的信用评分模型 https://cosx.org/2016/05/credit-scoring-model-in-internet-credit-reporting/

**Comment by XiangyunHuang** [comment]
支持向量机：是炒作还是哈利路亚（神器）

支持向量机和相关的核方法已经在数据挖掘任务中越来越流行，比如分类、回归和异常检测。本文的目的是从几何的角度对 SVM 提供一个直观的解释。我们基于分类问题来阐述 SVM 背后的基本概念，并从数据挖掘的角度来审视它们的优势和不足。尽管这个概述是不全面的，但它提供了必要的资源供那些有兴趣进一步探索 SVM 的人。

Support vector machines: hype or hallelujah? 2000, Kristin P. Bennett and Colin Campbell

https://doi.org/10.1145/380995.380999

**Comment by XiangyunHuang** [comment]
Get up to speed with Bayesian data analysis in R [Rasmus Bååth](http://www.sumsar.net/)

[Bayes tutorial - UseR 2019.pptx](https://github.com/XiangyunHuang/r-bloggers/files/3572345/Bayes.tutorial.-.UseR.2019.pptx)

涉及广告点击，促销活动因素影响，先验（看到数据之前对参数的认识）和后验（有数据之后，对参数的认识，学习模型、更新参数分布），以二项分布模拟广告点击场景，广告点击率的先验分布为均匀分布

rstanarm prophet CausalImpact 为贝叶斯推断工具

**Comment by XiangyunHuang** [comment]
- [ ] Why I don’t use the term “fixed and random effects” 为什么我不使用固定效应和随机效应这两个术语 Andrew Gelman  https://statmodeling.stat.columbia.edu/2005/01/25/why_i_dont_use/
- [ ] An introduction to Linear, Generalized, and Mixed/Multilevel models with R https://github.com/Pakillo/LM-GLM-GLMM-intro

**Comment by XiangyunHuang** [comment]
Three principles of data science: predictability, computability, and stability (PCS) Bin Yu, Karl Kumbier
数据科学三原则：可预测性P，可计算性C和稳定性S

PCS 从数据中取获取可再生的知识，指导科学假设的提出和试验设计。PCS 框架基于机器学习中的核心思想，将预测能力作为实践检验和评估计算，在数据收集、存储和算法设计。PC 比稳定性更加重要，它大大地扩展了传统统计的不确定性的考量。稳定性阐述结果关于选择/决策的变化机制，其横跨数据科学生命周期，包括问题形成，预处理、建模和建模前后的探索性分析（EDA）。进一步，我们提出 PCS 推断来探讨数据结果的稳定性和可识别性，模型在一致相对简单的情形下。我们比较了 PCS 推断和现存的方法，比如选择推断 selective inference，在高维稀疏线性模型及其不同参数的模拟设置下，就 ROC 曲线来说，我们的方法一致地优于其他方法。最后，我们提出 PCS 文档基于 R Markdown、IPython 或 Jupyter Notebook，公开可用，可重复的代码和在整个分析过程中人们的决策都备份下来了。在 Zenodo 上以基因案例研究的形式展示 PCS 工作流和文档。

https://arxiv.org/abs/1901.08152v3

数据和 Rmd 文档 [pcs_analyses.zip](https://github.com/XiangyunHuang/r-bloggers/files/3587498/pcs_analyses.zip)

**Comment by XiangyunHuang** [comment]
Interpretable machine learning: definitions, methods, and applications
W. James Murdoch, Chandan Singh, Karl Kumbier, Reza Abbasi-Asl, Bin Yu

机器学习模型在学习复杂模式用于预测方面已经取得巨大的成功。除了将模型用于预测外，解释模型学习到的东西的能力受到越来越多的关注。然而，这种日益增加的关注已经导致对可解释性概念的相当大的混淆。特别是，尚不清楚提出的各种解释方法是如何相关的，以及可以用什么共同概念来评估它们。

我们的目标是通过在机器学习的背景下定义可解释性并引入用于讨论解释的预测性，描述性，相关性（PDR）框架来解决这些问题。PDR框架为评估提供了三个总体需求：预测准确性，描述准确性和相关性，相对于人类受众判断相关性。此外，为了帮助管理大量的解释方法，我们引入了现有技术分类为基于模型和后期的类别，子组包括稀疏性，模块性和可模拟性。 为了演示从业者如何使用PDR框架来评估和理解解释，我们提供了许多真实的例子。 这些例子突出了人们在讨论中所起的作用往往被低估了解释性。 最后，基于我们的框架，我们讨论了现有方法的局限性和未来工作的方向。 我们希望这项工作能够提供一个共同的词汇表，使从业者和研究人员更容易讨论和选择各种解释方法。

https://arxiv.org/abs/1901.04592

还是统计研究人员看问题深刻，面对层出不穷的解释深度学习模型的图表，作者提出解释性的评估问题，Predictive, Descriptive, Relevant (PDR)

**Comment by XiangyunHuang** [comment]
2018 Artificial intelligence and statistics Bin YU, Karl KUMBIER

人工智能（AI）本质上是数据驱动的。 它要求在数据生成，算法开发和结果评估过程中通过人机协作应用统计概念。 本文讨论了如何通过这种方式来处理这种人机协作，人口统计概念P，兴趣问题Q，训练数据的代表性R和结果审查S（PQRS）。 PQRS工作流程提供了一个概念框架，用于将统计思想与人类输入整合到AI产品和研究中。 这些想法包括随机化和局部控制的实验设计原则以及稳定性原则，以获得算法和数据结果的可重复性和可解释性。我们将讨论在自动驾驶汽车，自动化医疗诊断以及作者协作研究的实例中使用这些原理。

https://doi.org//10.1631/FITEE.1700813

**Comment by XiangyunHuang** [comment]
- Appendix C: Computation in R and Stan http://www.stat.columbia.edu/~gelman/book/software.pdf

Andrew Gelman, John B. Carlin, Hal S. Stern, David B. Dunson, Aki Vehtari, and Donald B. Rubin (2013) Bayesian Data Analysis, Third Edition. Chapman & Hall/CRC Press.
free online: Appendix C: Computation in R and Stan

HMC/NUTS 算法的解释，配合代码

**Comment by XiangyunHuang** [comment]
- [ ] 书籍：利用 Python 做数据分析 https://github.com/BrambleXu/pydata-notebook

**Comment by XiangyunHuang** [comment]
- [ ] [GBDT 和 Xgboost 的区别](https://wepon.me/2016/05/07/XGBoost%E6%B5%85%E5%85%A5%E6%B5%85%E5%87%BA/)

**Comment by XiangyunHuang** [comment]
boosting 翻译过来是提升，实为集成的算法，组合多个分类器合成一个大分类器，获得更好的分类/回归性能，提升模型泛化能力

Boosting模型：lightGBM 算法原理 https://www.csuldw.com/2019/07/24/2019-07-24-an-introduction-tolightGBM-explained/

Boosting模型：XGBoost原理剖析  https://www.csuldw.com/2019/07/20/2019-07-20-xgboost-theory/

Boosting模型：GBDT原理介绍 https://www.csuldw.com/2019/07/12/2019-07-12-an-introduction-to-gbdt/

Gradient Tree Boosting Algorithm https://www.csuldw.com/2015/08/19/2015-08-19%20GBDT/

**Comment by XiangyunHuang** [comment]
- DALEX: explainers for complex predictive models https://arxiv.org/abs/1806.08915 https://github.com/ModelOriented/DALEX

**Comment by XiangyunHuang** [comment]
从线性模型到广义线性模型 (2)—参数估计、假设检验 https://cosx.org/2011/01/how-does-glm-generalize-lm-fit-and-test/ 广义线性模型的 IRWLS 算法

**Comment by XiangyunHuang** [comment]
[译作] 一些统计名词的新名字 https://cosx.org/2012/12/new-names-for-statistical-terms

**Comment by XiangyunHuang** [comment]
P2P 公司风控的难题

强监管下的数据获取问题
评测指标选择困难
模型在实际场景中的验证难题
最新技术（例如知识图谱）难以落地


P2P公司风控的优势

欺诈率高，不均衡问题没有那么严重
数据量小，对于大数据计算能力要求不高
金融公司核心业务，公司和行业重视度高

数据来源

线下门店数据
人口属性信息
工作信息
家庭信息
贷款信息

线上授权数据爬取
银行流水信息
通讯记录信息
社保信息

知识图谱数据
基于爬取用户数据构建的大规模知识图谱

算法模型
逻辑回归 （scikit-learn）
随机森林 ( scikit-learn )
GBDT (xgboost / lightGBM / catboost)
混合模型（逻辑回归 + GBDT）
图论模型

业内实践

某 P2P 公司官方数据： 逻辑回归、随机森林、GBDT、混合模型、深度学习 AUC 0.7x , ks 0.4x ，效果最好模型：混合模型

业内普遍水平： AUC 0.7x - 0.8x , ks 0.3x - 0.4x

最好效果是浅层模型，深度学习模型应用有待进一步发掘

图论模型通常是将图模式用作机器学习模型的特征


算法调试
Baseline : 逻辑回归
算法迭代：逻辑回归 -> 随机森林 -> xgboost / lightGBM
参数选择：Grid Search

金融科技公司修炼指南

不是所有基于规则的算法都是可解释的
不是所有的问题都可以用深度学习解决
不是所有的风控问题都需要精通金融业务
知识图谱刚刚起步，想清楚自己的意愿与能力是否匹配
不是技术好有想法就能在公司玩转一切
跟钱沾边儿的事儿风险也高，违法的事情尽量不参与
公司里鱼龙混杂很正常，把握住自己的方向


第一名的方案 https://www.kaggle.com/c/home-credit-default-risk/discussion/64821

**Comment by XiangyunHuang** [comment]
反向面试指南 
https://github.com/viraptor/reverse-interview
https://github.com/yifeikong/reverse-interview-zh

**Comment by XiangyunHuang** [comment]
算法/深度学习/NLP面试笔记
https://github.com/imhuay/Algorithm_Interview_Notes-Chinese

**Comment by XiangyunHuang** [comment]
# 算法工程师面试

## 深度学习 

### 模型评估方法

- Accuracy作为指标有哪些局限性？
- ROC曲线和PR曲线各是什么？
- 编程实现AUC的计算，并指出复杂度？
- AUC指标有什么特点？放缩结果对AUC是否有影响？
- 余弦距离与欧式距离有什么特点？

### 基本方法

- 如何划分训练集？如何选取验证集？
- 什么是偏差和方差？
- 什么是过拟合？深度学习解决过拟合的方法有哪
- 解决欠拟合的方法有哪些？
- 深度模型参数调整的一般方法论？

### 优化方法

- 简述了解的优化器，发展综述？
- 常用的损失函数有哪些？分别适用于什么场景？
- 梯度下降与拟牛顿法的异同？
- L1和L2正则分别有什么特点？为何L1稀疏？

### 深度学习基础

- 以一层隐层的神经网络，relu激活，MSE作为损失函数推导反向传播
- NN的权重参数能否初始化为0？
- 什么是梯度消失和梯度爆炸？
- 常用的激活函数，导数？
- relu的有优点？又有什么局限性？他们的系列改进方法是啥？
- sigmoid和tanh为什么会导致梯度消失？
- 一个隐层需要多少节点能实现包含n元输入的任意布尔函数？
- 多个隐层实现包含n元输入的任意布尔函数，需要多少节点和网络层？
- dropout为何能防止过拟合？
- dropout和BN 在前向传播和方向传播阶段的区别？

### CNN

- 给定卷积核的尺寸，特征图大小计算方法？
- 网络容量计算方法
- 共享参数有什么优点
- 常用的池化操作有哪些？有什么特点？
- CNN如何用于文本分类？
- resnet提出的背景和核心理论是？
- 空洞卷积是什么？有什么应用场景？

### RNN

- 简述RNN，LSTM，GRU的区别和联系
- 画出lstm的结构图，写出公式
- RNN的梯度消失问题？如何解决？
- lstm中是否可以用relu作为激活函数？
- lstm各个门分别使用什么激活函数？
- 简述seq2seq模型？
- seq2seq在解码时候有哪些方法？
- Attention机制是什么？



## 机器学习

### 基础

- 样本不均衡如何处理？
- 什么是生成模型什么是判别模型？

### 集成学习

- 集成学习的分类？有什么代表性的模型和方法？
- 如何从偏差和方差的角度解释bagging和boosting的原理？
- GBDT的原理？和Xgboost的区别联系？
- adaboost和gbdt的区别联系？

### 模型 

- 手推LR、Kmeans、SVM
- 简述ridge和lasson的区别和联系
- 树模型如何调参
- 树模型如何剪枝？
- 是否存一定存在参数，使得SVM的训练误差能到0
- 逻辑回归如何处理多分类？
- 决策树有哪些划分指标？区别与联系？
- 简述SVD和PCA的区别和联系？
- 如何使用梯度下降方法进行矩阵分解？
- LDA与PCA的区别与联系？

### 特征工程 

- 常用的特征筛选方法有哪些？
- 文本如何构造特征？
- 类别变量如何构造特征？
- 连续值变量如何构造特征？
- 哪些模型需要对特征进行归一化？
- 什么是组合特征？如何处理高维组合特征？

### 其他（分方向）

- word2vec的原理，glove的原理，fasttext的原理？

- cbow和skipgram如何选择？

- 了解elmo和bert吗？简述与word embedding的联系和区别

- 图像和文本和语音数据各有哪些数据增强方法？

- rcnn、fatse rcnn、fatser rcnn、mask rcnn的原理？ 

- 介绍resnet和GoogLeNet中的inception module的结构？ 

- 介绍yolo和ssd ？

- 介绍FM，FFM，deepFM，deepWide.

- 机器翻译如何解决oov？

- 等等

  

  

## 数据结构与算法

## 二叉树类

- 前中后的非递归？
- 层次遍历，之字遍历？
- 二叉树的序列化与反序列化
- 前中，后中遍历结果恢复二叉树
- 排序二叉树的序列化
- 二叉树的直径
- 二叉树的路径和为定值的路径
- 翻转、复制二叉树
- 排序二叉树转双向链表
- 判断二叉树是否相同？判断一棵树是不是另一棵树的子树？

### 搜索回溯

- 八皇后，全排列，组合
- 重复数字的排列，重复数字的组合
- 图的搜索
- A star

### 概率题 

- 用rand7构造rand10
- 轮盘赌
- 三角形内等概论随机选点
- 等等

### 动态规划

- 编辑距离
- 背包
- LCS

## 字符串

- 给定字符串是否符合正则表达式XXX
- 给定字符串是否是数字？
- KMP
- 超大数相加

## 数组、二分、快排系列

- 旋转数组查找
- 数组中的topk
- 旋转打印数组
- 行列递增矩阵查找数字
- 查找出现超过一半的数字
- 查找和为定值的两个数

### 链表

- 翻转链表
- 两个链表是否有交点
- 一个链表是否有环
- 链表的倒数k个节点
- 合并链表
- 逆序打印链表

### 排序

- 各种排序的复杂度最优最差平均
- 堆排序topk
- 快排的变化



## 海量数据题

- 海量日志的出现最多的10个字符串
- 10亿个1-10的数字排序
- trie 树
- hashmap
- 布隆过滤器
- bitmap

等等

**Comment by XiangyunHuang** [comment]
机器学习笔记（八）：面试中常问的问题
http://xiabanle.me/archives/94801474.html#more

**Comment by XiangyunHuang** [comment]
http://rcompanion.org/handbook/

Mangiafico, S.S. 2016. Summary and Analysis of Extension Program Evaluation in R, version 1.18.1. 

http://rcompanion.org/documents/RHandbookProgramEvaluation.pdf

详细介绍了各种参数和非参数检验 http://rcompanion.org/handbook/ 书籍长达 775 页
